File: matlab/examples/transmission_lines/CPW_Line.m
%
% Tutorials / CPW_Line
%
% Description at:
%
% Tested with
%  - Octave 3.8.1
%  - openEMS v0.0.32
%
% (C) 2014 Thorsten Liebig <thorsten.liebig@gmx.de>

close all
clear
clc

%% setup the simulation %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
physical_constants;
unit = 1e-6; % specify everything in um
CPW_length = 40000;
CPW_port_length = 10000;
CPW_width  = 1000;
CPW_gap    = 140;
substrate_thickness = 512;
substrate_width = 5000
substrate_epr = 3.66;
f_max = 10e9;
air_spacing = 7000

% use a finite line CPW waveguide
if 1
  feed_R = 50;
  pml_add_cells = [8 8 8 8 8 8];
  feed_shift_cells = 0;
  x_spacing = air_spacing;
else % or use a waveguide with start/end in a pml
  feed_R = inf; % CPW ends in a pml --> disable termination resistance 
  feed_shift_cells = 10; % CPW ends in an 8 cells thick pml --> shift feed 10 cells
  pml_add_cells = [0 0 8 8 8 8]; % do not add air-space in x-direction
  x_spacing = 0; % do not add air-space in x-direction
end

%% setup FDTD parameters & excitation function %%%%%%%%%%%%%%%%%%%%%%%%%%%%
FDTD = InitFDTD('EndCriteria', 1e-4);
FDTD = SetGaussExcite( FDTD, f_max/2, f_max/2 );
BC   = [2 2 2 2 2 2];
FDTD = SetBoundaryCond( FDTD, BC );

%% setup CSXCAD geometry & mesh %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
CSX = InitCSX();
resolution = c0/(f_max*sqrt(substrate_epr))/unit /30; % resolution of lambda/50
edge_res = 40;
mesh.x = SmoothMeshLines( [0 CPW_length/2 CPW_length/2+x_spacing], resolution, 1.5 ,0 );
mesh.x = unique(sort([-mesh.x mesh.x]));
mesh.y = SmoothMeshLines( [CPW_width/2+[-edge_res/3 +edge_res/3*2] CPW_gap+CPW_width/2+[-edge_res/3*2 +edge_res/3]], edge_res , 1.5 ,0);
mesh.y = SmoothMeshLines( [0 mesh.y], edge_res*2, 1.3 ,0);
mesh.y = SmoothMeshLines( [0 mesh.y substrate_width/2 substrate_width/2+air_spacing], resolution, 1.3 ,0);
mesh.y = unique(sort([-mesh.y mesh.y]));
mesh.z = SmoothMeshLines( [-air_spacing linspace(0,substrate_thickness,5) substrate_thickness+air_spacing], resolution );

mesh = AddPML(mesh, pml_add_cells);
CSX = DefineRectGrid( CSX, unit, mesh );

%% substrate
CSX = AddMaterial( CSX, 'RO4350B' );
CSX = SetMaterialProperty( CSX, 'RO4350B', 'Epsilon', substrate_epr );
start = [-CPW_length/2, -substrate_width/2, 0];
stop  = [+CPW_length/2, +substrate_width/2, substrate_thickness];
CSX = AddBox( CSX, 'RO4350B', 0, start, stop );

%%
CSX = AddMetal( CSX, 'CPW_PORT' );

%% CPW port, with the measurement plane at the end of each port
portstart = [ -CPW_length/2                , -CPW_width/2, substrate_thickness];
portstop  = [ -CPW_length/2+CPW_port_length,  CPW_width/2, substrate_thickness];
[CSX,port{1}] = AddCPWPort( CSX, 999, 1, 'CPW_PORT', portstart, portstop, CPW_gap, 'x', [0 1 0], 'ExcitePort', true, 'FeedShift', feed_shift_cells*resolution, 'MeasPlaneShift',  CPW_port_length, 'Feed_R', feed_R);

portstart = [ CPW_length/2                , -CPW_width/2, substrate_thickness];
portstop  = [ CPW_length/2-CPW_port_length,  CPW_width/2, substrate_thickness];
[CSX,port{2}] = AddCPWPort( CSX, 999, 2, 'CPW_PORT', portstart, portstop, CPW_gap, 'x', [0 1 0], 'MeasPlaneShift',  CPW_port_length, 'Feed_R', feed_R);

%% CPW
CSX = AddMetal( CSX, 'CPW');
start = [ -CPW_length/2+CPW_port_length, -CPW_width/2, substrate_thickness];
stop  = [ +CPW_length/2-CPW_port_length,  CPW_width/2, substrate_thickness];
CSX = AddBox(CSX, 'CPW', 999, start, stop);

%% CPW grounds
CSX = AddMetal( CSX, 'GND' );
start = [-CPW_length/2, -CPW_width/2-CPW_gap, substrate_thickness];
stop  = [+CPW_length/2, -substrate_width/2  , substrate_thickness];
CSX = AddBox(CSX, 'GND', 999, start, stop);

start = [-CPW_length/2, +CPW_width/2+CPW_gap, substrate_thickness];
stop  = [+CPW_length/2, +substrate_width/2  , substrate_thickness];
CSX = AddBox(CSX, 'GND', 999, start, stop);

%% write/show/run the openEMS compatible xml-file
Sim_Path = 'tmp';
Sim_CSX = 'CPW.xml';

[status, message, messageid] = rmdir( Sim_Path, 's' ); % clear previous directory
[status, message, messageid] = mkdir( Sim_Path ); % create empty simulation folder

WriteOpenEMS( [Sim_Path '/' Sim_CSX], FDTD, CSX );
CSXGeomPlot( [Sim_Path '/' Sim_CSX] );
RunOpenEMS( Sim_Path, Sim_CSX );

%% post-processing
close all
f = linspace( 1e6, f_max, 1601 );
port = calcPort( port, Sim_Path, f, 'RefImpedance', 50);

s11 = port{1}.uf.ref./ port{1}.uf.inc;
s21 = port{2}.uf.ref./ port{1}.uf.inc;

plot(f/1e9,20*log10(abs(s11)),'k-','LineWidth',2);
hold on;
grid on;
plot(f/1e9,20*log10(abs(s21)),'r--','LineWidth',2);
legend('S_{11}','S_{21}');
ylabel('S-Parameter (dB)','FontSize',12);
xlabel('frequency (GHz) \rightarrow','FontSize',12);

File: python/Tests/CPW_Line.py
# -*- coding: utf-8 -*-
"""
 Coplanar Waveguide (CPW) Line Test

 Verifies CPWPort setup and S-parameter extraction for a finite-length CPW
 transmission line on RO4350B substrate with 50-Ohm port terminations.

 Pass criteria:
   max(dB(S11)) < -20 dB   (low reflection)
   min(dB(S21)) > -0.5 dB  (near-lossless — substrate is modelled lossless)
   max(dB(S21)) < +0.01 dB (sign-error guard)

 Tested with
  - python 3.14
  - openEMS v0.0.36+

 (c) 2026 Thorsten Liebig <thorsten.liebig@gmx.de>

"""

import os, tempfile
import numpy as np

from CSXCAD  import ContinuousStructure
from openEMS import openEMS
from openEMS.physical_constants import *
from openEMS.ports import CPWPort

### Setup the simulation
Sim_Path = os.path.join(tempfile.gettempdir(), 'CPW_Line')

unit                = 1e-6   # drawing unit in um
CPW_length          = 40000
CPW_port_length     = 10000
CPW_width           = 1000
CPW_gap             = 140
substrate_thickness = 512
substrate_width     = 5000
substrate_epr       = 3.66
f_max               = 10e9
air_spacing         = 7000

feed_R = 50  # lumped port termination resistance

### Setup FDTD parameters & excitation
FDTD = openEMS(EndCriteria=1e-4)
FDTD.SetGaussExcite(f_max/2, f_max/2)
FDTD.SetBoundaryCond(['PMC']*6)

### Setup CSXCAD geometry & mesh
CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(unit)

resolution = C0 / (f_max * np.sqrt(substrate_epr)) / unit / 30
edge_res   = 40

# x-mesh: fine at port transitions, coarse in middle and beyond ports
mesh.AddLine('x', [0])
mesh.AddLine('x', [-CPW_length/2, -CPW_length/2 + CPW_port_length,
                    CPW_length/2 - CPW_port_length, CPW_length/2])
mesh.AddLine('x', [-CPW_length/2 - air_spacing, CPW_length/2 + air_spacing])
mesh.SmoothMeshLines('x', resolution, ratio=1.5)

# y-mesh: fine at CPW signal and gap edges, coarse elsewhere
third_mesh = np.array([-2/3, 1/3]) * edge_res
mesh.AddLine('y', [0])
mesh.AddLine('y',  CPW_width/2 + third_mesh)
mesh.AddLine('y',  CPW_width/2 + CPW_gap - third_mesh)
mesh.SmoothMeshLines('y', edge_res*1.5, ratio=1.5)
y_pos = mesh.GetLines('y')
mesh.AddLine('y', np.concatenate([-y_pos,
                                   [-substrate_width/2,  substrate_width/2],
                                   [-substrate_width/2 - air_spacing,
                                     substrate_width/2 + air_spacing]]))
mesh.SmoothMeshLines('y', resolution, ratio=1.3)

# z-mesh: fine inside substrate, coarse in air
mesh.AddLine('z', np.linspace(0, substrate_thickness, 5))
mesh.AddLine('z', [-air_spacing, substrate_thickness + air_spacing])
mesh.SmoothMeshLines('z', resolution)

### Substrate
substrate = CSX.AddMaterial('RO4350B', epsilon=substrate_epr)
start = [-CPW_length/2, -substrate_width/2, 0]
stop  = [ CPW_length/2,  substrate_width/2, substrate_thickness]
substrate.AddBox(start, stop)

### CPW ports (include the port metal)
cpw_port_metal = CSX.AddMetal('CPW_PORT')

portstart = [-CPW_length/2,                  -CPW_width/2, substrate_thickness]
portstop  = [-CPW_length/2 + CPW_port_length,  CPW_width/2, substrate_thickness]
port1 = CPWPort(CSX, 1, cpw_port_metal, portstart, portstop, 'x', 'y', CPW_gap,
                excite=1, priority=999,
                MeasPlaneShift=CPW_port_length, Feed_R=feed_R)

portstart = [ CPW_length/2,                  -CPW_width/2, substrate_thickness]
portstop  = [ CPW_length/2 - CPW_port_length,  CPW_width/2, substrate_thickness]
port2 = CPWPort(CSX, 2, cpw_port_metal, portstart, portstop, 'x', 'y', CPW_gap,
                priority=999, MeasPlaneShift=CPW_port_length, Feed_R=feed_R)

ports = [port1, port2]

### CPW centre conductor between the two ports
cpw = CSX.AddMetal('CPW')
start = [-CPW_length/2 + CPW_port_length, -CPW_width/2, substrate_thickness]
stop  = [ CPW_length/2 - CPW_port_length,  CPW_width/2, substrate_thickness]
cpw.AddBox(start, stop, priority=999)

### CPW ground planes (left and right of the gap)
gnd = CSX.AddMetal('GND')
start = [-CPW_length/2, -CPW_width/2 - CPW_gap, substrate_thickness]
stop  = [ CPW_length/2, -substrate_width/2,      substrate_thickness]
gnd.AddBox(start, stop, priority=999)

start = [-CPW_length/2,  CPW_width/2 + CPW_gap, substrate_thickness]
stop  = [ CPW_length/2,  substrate_width/2,      substrate_thickness]
gnd.AddBox(start, stop, priority=999)

if 1:  # debugging only
    CSX_file = os.path.join(Sim_Path, 'cpw_line.xml')
    if not os.path.exists(Sim_Path):
        os.mkdir(Sim_Path)
    CSX.Write2XML(CSX_file)
    from CSXCAD import AppCSXCAD_BIN
    os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

### Run the simulation
FDTD.Run(Sim_Path, cleanup=True)

### Post-processing
f = np.linspace(1e6, f_max, 1601)
for port in ports:
    port.CalcPort(Sim_Path, f, ref_impedance=50)

s11 = ports[0].uf_ref / ports[0].uf_inc
s21 = ports[1].uf_ref / ports[0].uf_inc

s11_dB = 20 * np.log10(np.abs(s11))
s21_dB = 20 * np.log10(np.abs(s21))

### Pass / fail checks
mask = f > 100e6
print(f'max(dB(S11)) = {np.max(s11_dB[mask]):.1f} dB')
print(f'min(dB(S21)) = {np.min(s21_dB[mask]):.1f} dB,  max(dB(S21)) = {np.max(s21_dB[mask]):.2f} dB')

assert np.max(s11_dB[mask]) < -20, \
    f'FAIL: max(dB(S11)) = {np.max(s11_dB[mask]):.1f} dB, expected < -20 dB'
assert np.min(s21_dB[mask]) > -0.5, \
    f'FAIL: min(dB(S21)) = {np.min(s21_dB[mask]):.1f} dB, expected > -0.5 dB'
assert np.max(s21_dB[mask]) < 0.05, \
    f'FAIL: max(dB(S21)) = {np.max(s21_dB[mask]):.2f} dB, expected < +0.05 dB (sign error?)'

print('PASS')

if 1:  # set to 1 for debugging plots
    import matplotlib.pyplot as plt

    fig, axis = plt.subplots(num='S-Parameters', tight_layout=True)
    axis.plot(f/1e9, s11_dB, 'k-',  linewidth=2, label='$S_{11}$')
    axis.plot(f/1e9, s21_dB, 'r--', linewidth=2, label='$S_{21}$')
    axis.grid()
    axis.set_xmargin(0)
    axis.set_xlabel('Frequency (GHz)')
    axis.set_ylabel('S-Parameter (dB)')
    axis.legend()

    plt.show()
